README.md
# NAME

Promise::ES6 - ES6-style promises in Perl

# SYNOPSIS

    my $promise = Promise::ES6->new( sub {
        my ($resolve_cr, $reject_cr) = @_;

        # ..
    } );

    my $promise2 = $promise->then( sub { .. }, sub { .. } );

    my $promise3 = $promise->catch( sub { .. } );

    my $promise4 = $promise->finally( sub { .. } );

    my $resolved = Promise::ES6->resolve(5);
    my $rejected = Promise::ES6->reject('nono');

    my $all_promise = Promise::ES6->all( \@promises );

    my $race_promise = Promise::ES6->race( \@promises );

# DESCRIPTION

This module provides a Perl implementation of [promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises), a useful pattern
for coordinating asynchronous tasks.

Unlike most other promise implementations on CPAN, this module
mimics ECMAScript 6’s [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)
interface. As the SYNOPSIS above shows, you can thus use patterns from
JavaScript in Perl with only minimal changes needed to accommodate language
syntax.

# INTERFACE NOTES

- Promise resolutions and rejections accept exactly one argument,
not a list.
- Unhandled rejections are reported via `warn()`. (See below
for details.)
- The Promises/A+ test suite avoids testing the case where an “executor”
function’s resolve callback itself receives another promise, e.g.:

        my $p = Promise::ES6->new( sub ($res) {
            $res->( Promise::ES6->resolve(123) );
        } );

    What will $p’s resolution value be? 123, or the promise that wraps it?

    This module favors conformity with the ES6 standard, which
    [indicates intent](https://www.ecma-international.org/ecma-262/6.0/#sec-promise-executor) that $p’s resolution value be 123.

# COMPATIBILITY

Right now this doesn’t interoperate directly with other promise
classes. If that’s something you want, make a feature request. For the
time being, of course, you can wrap one of this module’s promises in an
instance of whatever promise class you’re using, or vice-versa, to achieve
interoperability.

See [Promise::ES6::Future](https://metacpan.org/pod/Promise::ES6::Future) if you need to interact with [Future](https://metacpan.org/pod/Future).

# SPEED

By default this class is pure Perl; however, applications that manage
thousands of promises concurrently may want something faster.

Toward that goal, you can **EXPERIMENTALLY** load this class thus:

    use Promise::ES6 ( backend => 'XS' );

… to use [Promise::XS](https://metacpan.org/pod/Promise::XS) for most of the promise logic. This will confer a
significant speed advantage.

For maximum speed, though, use [Promise::XS](https://metacpan.org/pod/Promise::XS) directly.

# UNHANDLED REJECTIONS

As of version 0.05, unhandled rejections prompt a warning _only_ if one
of the following is true:

- 1) The unhandled rejection happens outside of the constructor.
- 2) The unhandled rejection happens via an uncaught exception
(even within the constructor).

# SYNCHRONOUS OPERATION

In JavaScript, the following …

    Promise.resolve().then( () => console.log(1) );
    console.log(2);

… will log `2` then `1` because JavaScript’s `then()` defers execution
of its callbacks until the end of the current iteration through JavaScript’s
event loop.

Perl, of course, has no built-in event loop. This module’s `then()` method,
thus, when called on a promise that is already
“settled” (i.e., not pending), will run the appropriate callback
_immediately_. That means that this:

    Promise::ES6->resolve(0)->then( sub { print 1 } );
    print 2;

… will print `12` instead of `21`.

This is an intentional divergence from
[the Promises/A+ specification](https://promisesaplus.com/#point-34).
An advantage of this design is that Promise::ES6 instances can abstract
over whether a given function works synchronously or asynchronously.

If you want a Promises/A+-compliant implementation, look at
[Promise::ES6::IOAsync](https://metacpan.org/pod/Promise::ES6::IOAsync), [Promise::ES6::Mojo](https://metacpan.org/pod/Promise::ES6::Mojo),
[Promise::ES6::AnyEvent](https://metacpan.org/pod/Promise::ES6::AnyEvent), or one of the alternatives
that that module’s documentation suggests.

# CANCELLATION

Promises have never provided a standardized solution for cancellation—i.e.,
aborting an in-process operation. So, if you need this functionality, you’ll
have to implement it yourself. Two ways of doing this are:

- Subclass Promise::ES6 and provide cancellation logic in that
subclass. See [DNS::Unbound::AsyncQuery](https://metacpan.org/pod/DNS::Unbound::AsyncQuery)’s implementation for an
example of this.
- Implement the cancellation on a request object that your
“promise-creator” also consumes. This is probably the more straightforward
approach but requires that there
be some object or ID besides the promise that uniquely identifies the action
to be canceled. See [Net::Curl::Promiser](https://metacpan.org/pod/Net::Curl::Promiser) for an example of this approach.

You’ll need to decide if it makes more sense for your application to leave
a canceled query in the “pending” state or to “settle” (i.e., resolve or
reject) it. All things being equal, I feel the first approach is the most
intuitive.

# MEMORY LEAKS

It’s easy to create inadvertent memory leaks using promises in Perl.
Here are a few “pointers” (heh) to bear in mind:

- As of version 0.07, any Promise::ES6 instances that are created while
`$Promise::ES6::DETECT_MEMORY_LEAKS` is set to a truthy value are
“leak-detect-enabled”, which means that if they survive until their original
process’s global destruction, a warning is triggered.
- If your application needs recursive promises (e.g., to poll
iteratively for completion of a task), the `current_sub` feature (i.e.,
`__SUB__`) may help you avoid memory leaks. In Perl versions that don’t
support this feature you can imitate it thus:

        use constant _has_current_sub => $^V ge v5.16.0;

        use if _has_current_sub(), feature => 'current_sub';

        my $cb;
        $cb = sub {
            my $current_sub = do {
                no strict 'subs';
                _has_current_sub() ? __SUB__ : eval '$cb';
            };
        }

    Of course, it’s better if you can avoid doing that. :)

- Garbage collection before Perl 5.18 seems to have been buggy.
If you work with such versions and end up chasing leaks,
try manually deleting as many references/closures as possible. See
`t/race_success.t` for a notated example.

    You may also (counterintuitively, IMO) find that this:

        my ($resolve, $reject);

        my $promise = Promise::ES6->new( sub { ($resolve, $reject) = @_ } );

        # … etc.

    … works better than:

        my $promise = Promise::ES6->new( sub {
            my ($resolve, $reject) = @_;

            # … etc.
        } );

# SEE ALSO

If you’re not sure of what promises are, there are several good
introductions to the topic. You might start with
[this one](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises).

[Promise::XS](https://metacpan.org/pod/Promise::XS) is a lot like this library but implemented mostly in XS for
speed.

Promise::ES6 serves much the same role as [Future](https://metacpan.org/pod/Future) but exposes
a standard, minimal, cross-language API rather than a proprietary (large) one.

CPAN contains a number of other modules that implement promises. A few
of note are:

- [Promise::XS](https://metacpan.org/pod/Promise::XS) and [AnyEvent::XSPromises](https://metacpan.org/pod/AnyEvent::XSPromises) - XS-powered promises.
Use one of these for maximum speed.
- [Promises](https://metacpan.org/pod/Promises) - Perhaps CPAN’s most widely-used promise implementation.
- [Mojo::Promise](https://metacpan.org/pod/Mojo::Promise) - Part of [Mojolicious](https://metacpan.org/pod/Mojolicious), tightly integrated to that
package.
- [Promise::Tiny](https://metacpan.org/pod/Promise::Tiny) - The forerunner to the present module. It
implements the same interface but has a number of bugs that proved hard
enough to fix that I felt a full rewrite was needed.

# LICENSE & COPYRIGHT

Copyright 2019-2020 Gasper Software Consulting.

This library is licensed under the same terms as Perl itself.


Changes
Revision history for Perl module Promise::ES6

0.19
- Rework tests
- Add experimental XS backend support via Promise::XS.
- Drop Perl 5.8 support. (5.8 can’t use Test2.)

0.18 Thu 16 Jan 2020
- Rejected promises no longer “repromise” their rejection values. Previously
  this module would attempt to settle a value given to a rejector callback
  in the same way as with a value given to a resolver callback; however,
  this behavior deviated from ES6 behavior. Since it was undocumented anyway,
  it’s removed here.
- Fix t/mojo.t to abort if given an unworkable Mojolicious version.

0.17 Sun 5 Jan 2020
- Optimization: remove extra call to isa(), and use $self rather than $_[0].
  (Thank you, bdraco!)

0.16 Tue 31 Dec 2019
- Optimization: reuse coderefs in all() and race().

0.15 Thu 19 Dec 2019
- BUG FIX: Make all() ignore rejections after the first one.
- Promise::ES6’s internal representation has changed to reduce memory usage.
  Any code that may depend on the former internals will likely be broken.
- Speed up implementation.
- Clarify intent when an executor function $resolve_cr is passed a promise.
- Make Mojo subclass require a sufficiently modern Mojo::IOLoop.
- Update Future.pm docs a bit.
- Refactor implementation to prepare for XS backend.

0.14 Wed 27 Nov 2019
- Fix test by requiring a reasonably recent Test::More.

0.13 Tue 26 Nov 2019
- Provide Future conversion logic.
- Flesh out Mojo.pm documentation.

0.12 Thu 21 Nov 2019
- Fix event-loop logic to run ->then() asynchronously.
- Remove left-in Data::Dumper.
- Add Mojolicious backend.

0.11 Fri 15 Nov 2019
- Fix memory leak in callback propagation.

0.10 Thu 14 Nov 2019
- Fix ::AnyEvent and ::IOAsync to pass values to callbacks.

0.09 Thu 14 Nov 2019
- Update documentation with discussions of synchronous operation and
  cancellation.
- Add Promise::ES6::AnyEvent and Promise::ES6::IOAsync to implement
  Promises/A+ support. (EXPERIMENTAL)
- Fix warnings in tests on old Perl versions.

0.08 Sun 22 Sep 2019
- Fix memory leaks in pre-5.18 perls, and proactively delete internal
  references.
- Update documentation about memory leaks.

0.07 Tue 17 Sep 2019
- Fix a number of memory leaks. (This necessitated some significant
  changes underneath.)
- Fix spurious unhandled-rejection warnings. These now happen at DESTROY
  time rather than right away. (I consider this too subtle of a change to
  be “breaking”.)
- Add more tests.

0.06 Thu 12 Sep 2019
- Issue #2: Make all() resolve when given an empty array reference.

0.05 Wed 31 Jul 2019
- Issue #1: Prevent warn() when creating a pre-rejected promise.

0.04 Mon 4 Mar 2019
- Fix missing test dependency. (doh!)

0.03 Sat 2 Mar 2019
- warn() on unhandled promise rejection.

0.02 Wed 13 Feb 2019
- Make tests Windows-compatible. (cf. RT #128491)
  No production changes.

0.01 Mon 11 Feb 2019
- Initial release


t/lib/EventTest.pm
package EventTest;

use strict;
use warnings;
use autodie;

use parent 'Test::Class::Tiny';

use Test::More;
use Test::FailWarnings -allow_deps => 1;

sub _FULL_BACKEND {
    return "Promise::ES6::" . $_[0]->_BACKEND();
}

sub T6_run {
    my ($self) = @_;

    my $class = ref $self;

  SKIP: {
        eval { $self->_REQUIRE(); 1 } or skip "$class: Backend isn’t available: $@", $self->num_tests();

        my $backend = $self->_BACKEND();
        require "Promise/ES6/$backend.pm";

        $self->_test_normal();
        $self->_test_die_in_constructor();
        $self->_test_resolve();
        $self->_test_reject();

        $self->_test_call_again_in_callback();
        $self->_test_die_in_then();
    }
}

sub _test_call_again_in_callback {
    my ($class) = @_;

    my @events;

    my $promise = $class->_FULL_BACKEND()->resolve(123123);

    my $promise_r = \$promise;

    my $big_promise = Promise::ES6->new( sub {
        my ($y, $n) = @_;

        $promise->then( sub {
            $$promise_r->then( sub { push @events, 1; $y->() } );

            push @events, 2;
        } );
    } );

#    my $cv = AnyEvent->condvar();
#
#    my $w = AnyEvent->idle( cb => sub {
#        if (@events == 2) {
#            $$promise_r = undef;
#            $cv->();
#        }
#    } );
#
#    $cv->recv();
    $class->_RESOLVE($big_promise);

    is( "@events", "2 1", 'then() from pre-resolved promise again' );
}

sub _test_normal {
    my ($class) = @_;

    my @things;

    my $promise = $class->_FULL_BACKEND()->new( sub {
        push @things, 'a';
        shift()->(123);
        push @things, 'b';
    } );

    push @things, 'c';

    $promise->then( sub { push @things, shift() } );

    push @things, 'e';

    $class->_RESOLVE($promise);

    push @things, 'f';

    is(
        "@things",
        'a b c e 123 f',
        'then() callback invoked asynchronously',
    );
}

sub _test_resolve {
    my ($class) = @_;

    my @things;

    my $promise = $class->_FULL_BACKEND()->resolve(123);

    push @things, 'c';

    $promise->then( sub { push @things, 'd' } );

    push @things, 'e';

    $class->_RESOLVE($promise);

    push @things, 'f';

    is(
        "@things",
        'c e d f',
        'then() callback invoked asynchronously',
    );
}

sub _test_reject {
    my ($class) = @_;

    my @things;

    my $promise = $class->_FULL_BACKEND()->reject(123);

    push @things, 'c';

    $promise->catch( sub { push @things, shift } );

    push @things, 'e';

    $class->_RESOLVE($promise);

    push @things, 'f';

    is(
        "@things",
        'c e 123 f',
        'catch() callback invoked asynchronously',
    );
}

sub _test_die_in_constructor {
    my ($class) = @_;

    my @things;

    my $promise = $class->_FULL_BACKEND()->new( sub {
        push @things, 'a';
        die "123\n";
        push @things, 'b';
    } );

    push @things, 'c';

    $promise->catch( sub { push @things, shift } );

    push @things, 'e';

    $class->_RESOLVE($promise);

    push @things, 'f';

    is(
        "@things",
        "a c e 123\n f",
        'catch() callback invoked asynchronously',
    );
}

sub _test_die_in_then {
    my ($class) = @_;

    my @things;

    my $promise = $class->_FULL_BACKEND()->resolve(123)->then( sub {
        die "123\n";
    } );

    push @things, 'c';

    $promise->catch( sub { push @things, shift } );

    push @things, 'e';

    $class->_RESOLVE($promise);

    push @things, 'f';

    is(
        "@things",
        "c e 123\n f",
        'catch() callback invoked asynchronously',
    );
}

1;


t/lib/BackendTest.pm
package BackendTest;

use strict;
use warnings;

use FindBin;

use Test::More;

sub run_modulinos {
    my @path = File::Spec->splitdir(__FILE__);
    pop @path;  # filename
    pop @path;  # up one dir
    push @path, 'modulinos';

    my $wildcard_path = File::Spec->catdir(@path, '*.t');

    my @modulino_nss;

    my @paths = glob $wildcard_path;

    _run_series(@paths);
}

sub _run_series {
    for my $path ( @_ ) {
        my @test_path = File::Spec->splitpath($path);
        my $filename = $test_path[-1];

        my $module_name_leaf = $filename;
        $module_name_leaf =~ s<\.t\z><> or die "weird modulino name: $module_name_leaf";

        my $ns = "t::$module_name_leaf";

        subtest $ns => sub {
            require $path;
            $ns->runtests();
        }
    }
}

1;


t/lib/MemoryCheck.pm
package MemoryCheck;

use strict;
use warnings;

use Test::More;

use Promise::ES6;

use Data::Dumper;

my $MASTER_PID = $$;

my $has_devel_gd;

BEGIN {
    if ( $has_devel_gd = eval { require Devel::GlobalDestruction; 1 } ) {
        diag "Devel::GlobalDestruction is available and loaded.";
    }
    else {
        diag "Devel::GlobalDestruction isn’t available.";
        if (!${^GLOBAL_PHASE}) {
            diag "XXX WARNING: This test won’t detect memory leaks.";
        }
    }

    # To ensure that we’ve loaded a backend:
    Promise::ES6->new( sub {} );

    my $destroy_cr = Promise::ES6->can('DESTROY');

    no warnings 'redefine';
    *Promise::ES6::DESTROY = sub {
        my $warn_gd = ($$ == $MASTER_PID);

        $warn_gd &&= $has_devel_gd ? Devel::GlobalDestruction::in_global_destruction() : (${^GLOBAL_PHASE} && ('DESTRUCT' eq ${^GLOBAL_PHASE}));

        if ($warn_gd) {
            print STDERR "XXX XXX XXX --- PID $$: DESTROYing Promise::ES6 at DESTRUCT time!$/";
            print STDERR Dumper(@_);

            # Avoid exit() so that we’ll see all possible problems.
            $? = 1;
        }

        return $destroy_cr->(@_) if $destroy_cr;
    };
}

1;


lib/Promise/ES6/Backend/PP.pm
package Promise::ES6;

#----------------------------------------------------------------------
# This module iS NOT a defined interface. Nothing to see here …
#----------------------------------------------------------------------

use strict;
use warnings;

use constant {

    # These aren’t actually defined.
    _RESOLUTION_CLASS => 'Promise::ES6::_RESOLUTION',
    _REJECTION_CLASS  => 'Promise::ES6::_REJECTION',
    _PENDING_CLASS    => 'Promise::ES6::_PENDING',
};

use constant {
    _PID_IDX         => 0,
    _CHILDREN_IDX    => 1,
    _VALUE_SR_IDX    => 2,
    _DETECT_LEAK_IDX => 3,
    _ON_RESOLVE_IDX  => 4,
    _ON_REJECT_IDX   => 5,
};

# "$value_sr" => $value_sr
our %_UNHANDLED_REJECTIONS;

sub _new {
    my ( $class, $cr ) = @_;

    die 'Need callback!' if !$cr;

    my $value;
    my $value_sr = bless \$value, _PENDING_CLASS();

    my @children;

    my $self = bless [
        $$,
        \@children,
        $value_sr,
        $Promise::ES6::DETECT_MEMORY_LEAKS,
    ], $class;

    my $suppress_unhandled_rejection_warning = 1;

    # NB: These MUST NOT refer to $self, or else we can get memory leaks
    # depending on how $resolver and $rejector are used.
    my $resolver = sub {
        $$value_sr = $_[0];
        bless $value_sr, _RESOLUTION_CLASS();

        # NB: UNIVERSAL::isa() is used in order to avoid an eval {}.
        # It is acknowledged that many Perl experts strongly discourage
        # use of this technique.
        if ( UNIVERSAL::isa( $$value_sr, __PACKAGE__ ) ) {
            _repromise( $value_sr, \@children, $value_sr );
        }
        elsif (@children) {
            $_->_settle($value_sr) for splice @children;
        }
    };

    my $rejecter = sub {
        $$value_sr = $_[0];
        bless $value_sr, _REJECTION_CLASS();

        if ( !$suppress_unhandled_rejection_warning ) {
            $_UNHANDLED_REJECTIONS{$value_sr} = $value_sr;
        }

        # We do not repromise rejections. Whatever is in $$value_sr
        # is literally what rejection callbacks receive.
        if (@children) {
            $_->_settle($value_sr) for splice @children;
        }
    };

    local $@;
    if ( !eval { $cr->( $resolver, $rejecter ); 1 } ) {
        $$value_sr = $@;
        bless $value_sr, _REJECTION_CLASS();

        $_UNHANDLED_REJECTIONS{$value_sr} = $value_sr;
    }

    $suppress_unhandled_rejection_warning = 0;

    return $self;
}

sub then {
    my ( $self, $on_resolve, $on_reject ) = @_;

    my $value_sr = bless( \do { my $v }, _PENDING_CLASS() );

    my $new = bless [
        $$,
        [],
        $value_sr,
        $Promise::ES6::DETECT_MEMORY_LEAKS,
        $on_resolve,
        $on_reject,
      ],
      ref($self);

    if ( _PENDING_CLASS eq ref $self->[_VALUE_SR_IDX] ) {
        push @{ $self->[_CHILDREN_IDX] }, $new;
    }
    else {

        # $self might already be settled, in which case we immediately
        # settle the $new promise as well.

        $new->_settle( $self->[_VALUE_SR_IDX] );
    }

    return $new;
}

#----------------------------------------------------------------------

sub _repromise {
    my ( $value_sr, $children_ar, $repromise_value_sr ) = @_;
    $$repromise_value_sr->then(
        sub {
            $$value_sr = $_[0];
            bless $value_sr, _RESOLUTION_CLASS;
            $_->_settle($value_sr) for splice @$children_ar;
        },
        sub {
            $$value_sr = $_[0];
            bless $value_sr, _REJECTION_CLASS;
            $_->_settle($value_sr) for splice @$children_ar;
        },
    );
    return;

}

# It’s gainfully faster to inline this:
#sub _is_completed {
#    return (_PENDING_CLASS ne ref $_[0][ _VALUE_SR_IDX ]);
#}

my $settle_is_rejection;

# This method *only* runs when $self is “settled”:
sub _settle {
    my ( $self, $final_value_sr ) = @_;

    die "$self already settled!" if _PENDING_CLASS ne ref $_[0][_VALUE_SR_IDX];

    $settle_is_rejection = _REJECTION_CLASS eq ref $final_value_sr;

    # A promise that new() created won’t have on-settle callbacks,
    # but a promise that came from then/catch/finally will.
    # It’s a good idea to delete the callbacks in order to trigger garbage
    # collection as soon and as reliably as possible. It’s safe to do so
    # because _settle() is only called once.
    my $callback = $self->[ $settle_is_rejection ? _ON_REJECT_IDX : _ON_RESOLVE_IDX ];

    @{$self}[ _ON_RESOLVE_IDX, _ON_REJECT_IDX ] = ();

    # Only needed when $settle_is_rejection, but the check would be more
    # expensive than just always deleting. So, hey.
    delete $_UNHANDLED_REJECTIONS{$final_value_sr};

    my $value_sr_contents_is_package = 1;

    if ($callback) {

        # This is the block that runs for promises that were created by a
        # call to then() that assigned a handler for the state that
        # $final_value_sr indicates (i.e., resolved or rejected).

        my ($new_value);

        local $@;

        if ( eval { $new_value = $callback->($$final_value_sr); 1 } ) {

            # The callback succeeded. If $new_value is not itself a promise,
            # then $self is now resolved. (Yay!) Note that this is true
            # even if $final_value_sr indicates a rejection: in this case, we’ve
            # just run a successful “catch” block, so resolution is correct.

            # If $new_value IS a promise, though, then we have to wait.
            if ( !UNIVERSAL::isa( $new_value, __PACKAGE__ ) ) {
                $value_sr_contents_is_package = 0;
                bless $self->[_VALUE_SR_IDX], _RESOLUTION_CLASS();
            }
        }
        else {

            # The callback errored, which means $self is now rejected.

            $new_value                    = $@;
            $value_sr_contents_is_package = 0;

            bless $self->[_VALUE_SR_IDX], _REJECTION_CLASS();
            $_UNHANDLED_REJECTIONS{ $self->[_VALUE_SR_IDX] } = $self->[_VALUE_SR_IDX];
        }

        ${ $self->[_VALUE_SR_IDX] } = $new_value;
    }
    else {

        # There was no handler from then(), so whatever state $final_value_sr
        # indicates # (i.e., resolution or rejection) is now $self’s state
        # as well.

        bless $self->[_VALUE_SR_IDX], ref($final_value_sr);
        ${ $self->[_VALUE_SR_IDX] } = $$final_value_sr;
        $value_sr_contents_is_package = UNIVERSAL::isa( $$final_value_sr, __PACKAGE__ );

        if ($settle_is_rejection) {
            $_UNHANDLED_REJECTIONS{ $self->[_VALUE_SR_IDX] } = $self->[_VALUE_SR_IDX];
        }
    }

    if ($value_sr_contents_is_package) {
        return _repromise( @{$self}[ _VALUE_SR_IDX, _CHILDREN_IDX, _VALUE_SR_IDX ] );
    }

    if ( @{ $self->[_CHILDREN_IDX] } ) {
        $_->_settle( $self->[_VALUE_SR_IDX] ) for splice @{ $self->[_CHILDREN_IDX] };
    }

    return;
}

sub DESTROY {
    return if $$ != $_[0][_PID_IDX];

    if ( $_[0][_DETECT_LEAK_IDX] && ${^GLOBAL_PHASE} && ${^GLOBAL_PHASE} eq 'DESTRUCT' ) {
        warn( ( '=' x 70 ) . "\n" . 'XXXXXX - ' . ref( $_[0] ) . " survived until global destruction; memory leak likely!\n" . ( "=" x 70 ) . "\n" );
    }

    if ( my $promise_value_sr = $_[0][_VALUE_SR_IDX] ) {
        if ( my $value_sr = delete $_UNHANDLED_REJECTIONS{$promise_value_sr} ) {
            my $ref = ref $_[0];
            warn "$ref: Unhandled rejection: $$value_sr";
        }
    }
}

1;


lib/Promise/ES6.pm
package Promise::ES6;

use strict;
use warnings;

our $VERSION = '0.18';

=encoding utf-8

=head1 NAME

Promise::ES6 - ES6-style promises in Perl

=head1 SYNOPSIS

    my $promise = Promise::ES6->new( sub {
        my ($resolve_cr, $reject_cr) = @_;

        # ..
    } );

    my $promise2 = $promise->then( sub { .. }, sub { .. } );

    my $promise3 = $promise->catch( sub { .. } );

    my $promise4 = $promise->finally( sub { .. } );

    my $resolved = Promise::ES6->resolve(5);
    my $rejected = Promise::ES6->reject('nono');

    my $all_promise = Promise::ES6->all( \@promises );

    my $race_promise = Promise::ES6->race( \@promises );

=head1 DESCRIPTION

This module provides a Perl implementation of L<promises|https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises>, a useful pattern
for coordinating asynchronous tasks.

Unlike most other promise implementations on CPAN, this module
mimics ECMAScript 6’s L<Promise|https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise>
interface. As the SYNOPSIS above shows, you can thus use patterns from
JavaScript in Perl with only minimal changes needed to accommodate language
syntax.

=head1 INTERFACE NOTES

=over

=item * Promise resolutions and rejections accept exactly one argument,
not a list.

=item * Unhandled rejections are reported via C<warn()>. (See below
for details.)

=item * The Promises/A+ test suite avoids testing the case where an “executor”
function’s resolve callback itself receives another promise, e.g.:

    my $p = Promise::ES6->new( sub ($res) {
        $res->( Promise::ES6->resolve(123) );
    } );

What will $p’s resolution value be? 123, or the promise that wraps it?

This module favors conformity with the ES6 standard, which
L<indicates intent|https://www.ecma-international.org/ecma-262/6.0/#sec-promise-executor> that $p’s resolution value be 123.

=back

=head1 COMPATIBILITY

Right now this doesn’t interoperate directly with other promise
classes. If that’s something you want, make a feature request. For the
time being, of course, you can wrap one of this module’s promises in an
instance of whatever promise class you’re using, or vice-versa, to achieve
interoperability.

See L<Promise::ES6::Future> if you need to interact with L<Future>.

=head1 SPEED

By default this class is pure Perl; however, applications that manage
thousands of promises concurrently may want something faster.

Toward that goal, you can B<EXPERIMENTALLY> load this class thus:

    use Promise::ES6 ( backend => 'XS' );

… to use L<Promise::XS> for most of the promise logic. This will confer a
significant speed advantage.

For maximum speed, though, use L<Promise::XS> directly.

=head1 UNHANDLED REJECTIONS

As of version 0.05, unhandled rejections prompt a warning I<only> if one
of the following is true:

=over

=item 1) The unhandled rejection happens outside of the constructor.

=item 2) The unhandled rejection happens via an uncaught exception
(even within the constructor).

=back

=head1 SYNCHRONOUS OPERATION

In JavaScript, the following …

    Promise.resolve().then( () => console.log(1) );
    console.log(2);

… will log C<2> then C<1> because JavaScript’s C<then()> defers execution
of its callbacks until the end of the current iteration through JavaScript’s
event loop.

Perl, of course, has no built-in event loop. This module’s C<then()> method,
thus, when called on a promise that is already
“settled” (i.e., not pending), will run the appropriate callback
I<immediately>. That means that this:

    Promise::ES6->resolve(0)->then( sub { print 1 } );
    print 2;

… will print C<12> instead of C<21>.

This is an intentional divergence from
L<the Promises/A+ specification|https://promisesaplus.com/#point-34>.
An advantage of this design is that Promise::ES6 instances can abstract
over whether a given function works synchronously or asynchronously.

If you want a Promises/A+-compliant implementation, look at
L<Promise::ES6::IOAsync>, L<Promise::ES6::Mojo>,
L<Promise::ES6::AnyEvent>, or one of the alternatives
that that module’s documentation suggests.

=head1 CANCELLATION

Promises have never provided a standardized solution for cancellation—i.e.,
aborting an in-process operation. So, if you need this functionality, you’ll
have to implement it yourself. Two ways of doing this are:

=over

=item * Subclass Promise::ES6 and provide cancellation logic in that
subclass. See L<DNS::Unbound::AsyncQuery>’s implementation for an
example of this.

=item * Implement the cancellation on a request object that your
“promise-creator” also consumes. This is probably the more straightforward
approach but requires that there
be some object or ID besides the promise that uniquely identifies the action
to be canceled. See L<Net::Curl::Promiser> for an example of this approach.

=back

You’ll need to decide if it makes more sense for your application to leave
a canceled query in the “pending” state or to “settle” (i.e., resolve or
reject) it. All things being equal, I feel the first approach is the most
intuitive.

=head1 MEMORY LEAKS

It’s easy to create inadvertent memory leaks using promises in Perl.
Here are a few “pointers” (heh) to bear in mind:

=over

=item * As of version 0.07, any Promise::ES6 instances that are created while
C<$Promise::ES6::DETECT_MEMORY_LEAKS> is set to a truthy value are
“leak-detect-enabled”, which means that if they survive until their original
process’s global destruction, a warning is triggered.

=item * If your application needs recursive promises (e.g., to poll
iteratively for completion of a task), the C<current_sub> feature (i.e.,
C<__SUB__>) may help you avoid memory leaks. In Perl versions that don’t
support this feature you can imitate it thus:

    use constant _has_current_sub => $^V ge v5.16.0;

    use if _has_current_sub(), feature => 'current_sub';

    my $cb;
    $cb = sub {
        my $current_sub = do {
            no strict 'subs';
            _has_current_sub() ? __SUB__ : eval '$cb';
        };
    }

Of course, it’s better if you can avoid doing that. :)

=item * Garbage collection before Perl 5.18 seems to have been buggy.
If you work with such versions and end up chasing leaks,
try manually deleting as many references/closures as possible. See
F<t/race_success.t> for a notated example.

You may also (counterintuitively, IMO) find that this:

    my ($resolve, $reject);

    my $promise = Promise::ES6->new( sub { ($resolve, $reject) = @_ } );

    # … etc.

… works better than:

    my $promise = Promise::ES6->new( sub {
        my ($resolve, $reject) = @_;

        # … etc.
    } );

=back

=head1 SEE ALSO

If you’re not sure of what promises are, there are several good
introductions to the topic. You might start with
L<this one|https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises>.

L<Promise::XS> is a lot like this library but implemented mostly in XS for
speed.

Promise::ES6 serves much the same role as L<Future> but exposes
a standard, minimal, cross-language API rather than a proprietary (large) one.

CPAN contains a number of other modules that implement promises. A few
of note are:

=over

=item * L<Promise::XS> and L<AnyEvent::XSPromises> - XS-powered promises.
Use one of these for maximum speed.

=item * L<Promises> - Perhaps CPAN’s most widely-used promise implementation.

=item * L<Mojo::Promise> - Part of L<Mojolicious>, tightly integrated to that
package.

=item * L<Promise::Tiny> - The forerunner to the present module. It
implements the same interface but has a number of bugs that proved hard
enough to fix that I felt a full rewrite was needed.

=back

=head1 LICENSE & COPYRIGHT

Copyright 2019-2020 Gasper Software Consulting.

This library is licensed under the same terms as Perl itself.

=cut

#----------------------------------------------------------------------

our $DETECT_MEMORY_LEAKS;

sub new {
    _load_backend();

    *new = *_new;

    return _new(@_);
}

sub import {
    my %opts = @_[ 1 .. $#_ ];

    _load_backend( $opts{'backend'} );
}

sub catch { $_[0]->then( undef, $_[1] ) }

sub finally { $_[0]->then( $_[1], $_[1] ) }

sub resolve {
    my ( $class, $value ) = @_;

    $class->new( sub { $_[0]->($value) } );
}

sub reject {
    my ( $class, $reason ) = @_;

    $class->new( sub { $_[1]->($reason) } );
}

sub all {
    my ( $class, $iterable ) = @_;
    my @promises = map { UNIVERSAL::isa( $_, __PACKAGE__ ) ? $_ : $class->resolve($_) } @$iterable;

    my @values;

    return $class->new(
        sub {
            my ( $resolve, $reject ) = @_;
            my $unresolved_size = scalar(@promises);

            my $settled;

            if ($unresolved_size) {
                my $p = 0;

                my $on_reject_cr = sub {

                    # Needed because we might get multiple failures:
                    return if $settled;

                    $settled = 1;
                    $reject->(@_);
                };

                for my $promise (@promises) {
                    last if $settled;

                    my $p = $p++;

                    $promise->then(
                        sub {
                            return if $settled;

                            $values[$p] = $_[0];

                            $unresolved_size--;
                            return if $unresolved_size > 0;

                            $settled = 1;
                            $resolve->( \@values );
                        },
                        $on_reject_cr,
                    );
                }
            }
            else {
                $resolve->( [] );
            }
        }
    );
}

sub race {
    my ( $class, $iterable ) = @_;
    my @promises = map { UNIVERSAL::isa( $_, __PACKAGE__ ) ? $_ : $class->resolve($_) } @$iterable;

    my ( $resolve, $reject );

    # Perl 5.16 and earlier leak memory when the callbacks are handled
    # inside the closure here.
    my $new = $class->new(
        sub {
            ( $resolve, $reject ) = @_;
        }
    );

    my $is_done;

    my $on_resolve_cr = sub {
        return if $is_done;
        $is_done = 1;

        $resolve->( $_[0] );

        # Proactively eliminate references:
        $resolve = $reject = undef;
    };

    my $on_reject_cr = sub {
        return if $is_done;
        $is_done = 1;

        $reject->( $_[0] );

        # Proactively eliminate references:
        $resolve = $reject = undef;
    };

    for my $promise (@promises) {
        last if $is_done;

        $promise->then( $on_resolve_cr, $on_reject_cr );
    }

    return $new;
}

#----------------------------------------------------------------------

our $BACKEND;

sub _load_backend {
    my $want_backend = shift;

    if ($BACKEND) {
        if ($want_backend && $want_backend ne $BACKEND) {
            warn( __PACKAGE__ .  ": Requested “$want_backend” backend after “$BACKEND” is already loaded!\n" );
        }

        return;
    }

    if ($want_backend) {
        if ($want_backend eq 'XS') {
            require Promise::ES6::Backend::XS;
            $BACKEND = 'XS';
            return;
        }
        elsif ($want_backend ne 'PP') {
            die( __PACKAGE__ . ": Unknown backend ($want_backend); try “XS” or “PP”.\n");
        }
    }

    require Promise::ES6::Backend::PP;
    $BACKEND = 'PP';
}

1;
